base/pointer.jl

```diff
@@ -96,34 +96,168 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
             p::Ptr{T}, dims::NTuple{N,<:Integer}; own::Bool = false) where {T,N} =
     unsafe_wrap(Atype, p, convert(Tuple{Vararg{Int}}, dims), own = own)
 
+const ATOMIC_PTR_DOCS = """
+
+"""
+
 """
     unsafe_load(p::Ptr{T}, i::Integer=1)
+    unsafe_load(p::Ptr{T}, order::Symbol)
+    unsafe_load(p::Ptr{T}, i::Integer, order::Symbol)
 
 Load a value of type `T` from the address of the `i`th element (1-indexed) starting at `p`.
-This is equivalent to the C expression `p[i-1]`.
+This is equivalent to the C expression `p[i-1]`. Optionally, an ordering can be provided.
 
 The `unsafe` prefix on this function indicates that no validation is performed on the
 pointer `p` to ensure that it is valid. Like C, the programmer is responsible for ensuring
 that referenced memory is not freed or garbage collected while invoking this function.
 Incorrect usage may segfault your program or return garbage answers. Unlike C, dereferencing
 memory region allocated as different type may be valid provided that the types are compatible.
+
+This is compatible with loading pointers declared with `_Atomic` and `std::atomic`
+type in C11 and C++23 respectively. An error may be thrown if there is not support for
+atomically loading the Julia type `T`.
+
+!!! compat "Julia 1.10"
+     The `order` argument is available as of Julia 1.10.
 """
 unsafe_load(p::Ptr, i::Integer=1) = pointerref(p, Int(i), 1)
+unsafe_load(p::Ptr, order::Symbol) = atomic_pointerref(p, order)
+function unsafe_load(p::Ptr, i::Integer, order::Symbol)
+    unsafe_load(p + (aligned_sizeof(eltype(p)) * (Int(i) - 1)), order)
```

Comment:
We need to use a definition of `aligned_sizeof` that actually corresponds to what the intrinsics expect.

```julia
function elsize(::Type{Ptr{T}}) where T
    # this only must return something valid for values which satisfy is_valid_intrinsic_elptr(T),
    # which includes Any and most concrete datatypes
    T === Any && return sizeof(Ptr{Any})
    T isa DataType || sizeof(Any) # throws
    return LLT_ALIGN(Core.sizeof(T), datatype_alignment(T))
end
```

```suggestion
    unsafe_load(p + (elsize(typeof(p)) * (Int(i) - 1)), order)
```